src/sentry/static/sentry/app/views/settings/components/settingsProjectItem.jsx

```diff
@@ -0,0 +1,99 @@
+import styled from 'react-emotion';
+
+import React from 'react';
+
+import createReactClass from 'create-react-class';
+
+import {update} from '../../../actionCreators/projects';
+import ApiMixin from '../../../mixins/apiMixin';
+import TooltipMixin from '../../../mixins/tooltip';
+import Link from '../../../components/link';
+import ProjectLabel from '../../../components/projectLabel';
+import SentryTypes from '../../../proptypes';
+
+const StyledH5 = styled('h5')`
+  margin: 0;
+  a {
+    color: ${t => t.theme.gray3};
+  }
+`;
+
+const ProjectItem = createReactClass({
+  displayName: 'ProjectItem',
+
+  propTypes: {
+    project: SentryTypes.Project,
+    organization: SentryTypes.Organization,
+  },
+
+  mixins: [
+    ApiMixin,
+    TooltipMixin(function() {
+      return {
+        selector: '.tip',
+        title: function(instance) {
+          return this.getAttribute('data-isbookmarked') === 'true'
+            ? 'Remove from bookmarks'
+            : 'Add to bookmarks';
+        },
+      };
+    }),
+  ],
+
+  getInitialState() {
+    return {
+      bookmarked: null,
+    };
+  },
+
+  componentWillReceiveProps(nextProps) {
+    // Local bookmarked state should be unset when the project data changes
+    // Local state is used for optimistic UI update
+    if (nextProps.project.isBookmarked !== this.props.project.isBookmarked) {
+      this.setState({bookmarked: null});
```

Comment:
I was initially confused by this logic because we're relying on behavior when `bookmarked == null`. I think it'd be more clear if `state.bookmarked` was the "source of truth" (instead of `state.bookmarked || project.isBookmarked`)